provider/pvutil.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2016-2023, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package provider
+
+import (
+	"github.com/pulumi/pulumi-terraform-bridge/v3/unstable/propertyvalue"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+)
+
+// Assits building transformations on PropertyValue that do not want to deal with Secret, Computed,
+// or Output values.
+//
+// The usage pattern looks like this:
+//
+//	composePropertyValue[T](func (c *pvComposer) (resource.PropertyValue, T, error) {
+//	      x, _ := c.Simplify(pv1)
+//	      y, _ := c.SimplifyPropertyMap(pv2) // etc
+//	      return resource.NewArrayProperty([]resource.PropertyValue{x, y}), result, nil
+//	})
+//
+// User code accessing values that passed through Simplify is guaranteed to never observe Secret,
+// Computed, or Output values. All the metadata bits about these is floated to top-level and
+// re-applied to the value the user code receives out of composePropertyValue.
```

Comment:
I think this is fine. It means that `tags: { "foo": "bar", "secret": [secret] }` will come out as `[secret]`. I doubt that this will be a problem in practice. I don't want to block merging on it.

Reply:
Hmm you're right. We're losing precision. Unlike what happens in the restricted CheckConfigure, there are true nested secrets here.  That's bad especially when merging is not even involved, like empty provider tags + secret resource tags. Hm.. I may do a quick fix to this by keeping a track of secret string hashes and re-secreting matching in the output.